homeassistant/helpers/collection.py

```diff
@@ -312,29 +327,53 @@ def sync_entity_lifecycle(
 ) -> None:
     """Map a collection to an entity component."""
     entities = {}
+    ent_reg = entity_registry.async_get(hass)
 
-    async def _collection_changed(change_type: str, item_id: str, config: dict) -> None:
-        """Handle a collection change."""
-        if change_type == CHANGE_ADDED:
-            entity = create_entity(config)
-            await entity_component.async_add_entities([entity])
-            entities[item_id] = entity
-            return
-
-        if change_type == CHANGE_REMOVED:
-            ent_reg = await entity_registry.async_get_registry(hass)
-            ent_to_remove = ent_reg.async_get_entity_id(domain, platform, item_id)
-            if ent_to_remove is not None:
-                ent_reg.async_remove(ent_to_remove)
-            else:
-                await entities[item_id].async_remove(force_remove=True)
-            entities.pop(item_id)
-            return
+    async def _add_entity(change_set: CollectionChangeSet) -> Entity:
+        entities[change_set.item_id] = create_entity(change_set.item)
+        return entities[change_set.item_id]
 
-        # CHANGE_UPDATED
-        await entities[item_id].async_update_config(config)  # type: ignore
+    async def _remove_entity(change_set: CollectionChangeSet) -> None:
+        ent_to_remove = ent_reg.async_get_entity_id(
+            domain, platform, change_set.item_id
+        )
+        if ent_to_remove is not None:
+            ent_reg.async_remove(ent_to_remove)
+        else:
+            await entities[change_set.item_id].async_remove(force_remove=True)
+        entities.pop(change_set.item_id)
+
+    async def _update_entity(change_set: CollectionChangeSet) -> None:
+        await entities[change_set.item_id].async_update_config(change_set.item)  # type: ignore
+
+    _func_map = {
+        CHANGE_ADDED: _add_entity,
+        CHANGE_REMOVED: _remove_entity,
+        CHANGE_UPDATED: _update_entity,
+    }
+
+    async def _collection_changed(change_sets: Iterable[CollectionChangeSet]) -> None:
+        """Handle a collection change."""
+        # Create a new bucket every time we have a different change type
+        # to ensure operations happen in order. We only group
+        # the same change type.
+        task_buckets: list = []
+        previous_change_type = None
```

Comment:
This is never set ?

Reply:
I think that you could just use [`itertools.groupby`](https://docs.python.org/3/library/itertools.html#itertools.groupby) for this

Reply:
```python
from itertools import groupby

for key, grouped in groupby(change_sets, lambda change_set: change_set.change_type):
    await asyncio.gather(*[_func_map[change_set.change_type](change_set) for change_set in change_sets])
```